message ModuleIntro {
  required string long_name = 1; // for management
  required string short_name = 2; // for name prefixing
}

message BindingRequest {
  required string bind_info = 1;
}

message BindingResponse {
  required bool success = 1;
  optional string diagnostic = 2;
}

message UnbindRequest {
}

message UserStatus {
  optional bool online_status = 2;
}

message UserEvent {
  required string username = 1;
  optional string plugin_id = 2; // for foreign users
  //oneof {
  optional UserStatus user_status = 101;
  optional string chat_message = 102;
  //}
}

message GroupStatus {
  optional string topic = 1;
}

message GroupMessage {
  required uint32 group_id = 1;
  //oneof {
  optional BindingRequest binding_request = 101;
  optional BindingResponse binding_response = 102;
  optional UnbindRequest unbind_request = 103;
  optional GroupStatus group_status_change = 104;
  optional UserEvent user_event = 105;
  //}
}

